### ISSUE TITLE
0.10.0: updateCameraPosition stops previous updateCameraPosition (Android)

### ISSUE THREAD

**Issue by westnordost** [tangrams/tangram-es#1962]
Consider this very common use case: You have a map which centers on the current GPS position and orients itself towards the smartphone's orientation toward North.

In **0.9.6**, the code to do that was:
* on new GPS location received: `mapController.setPositionEased(position,500);`
* on compass sensor event received: `mapController.setRotation(rotation);`

Now, in **0.10.0**, I have to do:
* on new GPS location received: `mapController.updateCameraPosition(CameraUpdateFactory.setPosition(pos), 500);`
* on compass sensor event received: `mapController.updateCameraPosition(CameraUpdateFactory.setRotation(mapRotation));`

However, there is a problem: It seems that the animation towards the new position triggered by the new GPS location is interrupted and cancelled by the second call to `updateCameraPosition` by the compass sensor thread. (This happens at various places, the GPS sensor VS the orientation sensor is just the best example for it because it is such a common use case.)

Animating single camera properties was not a problem in 0.9.6, as far as I know. I do not find it understandable that 0.10.0 apparently does not have this feature anymore. Shouldn't the new verbose syntax make it more versatile rather than less?

**Comment by matteblair** [comment]
You can still achieve the behavior you want, if I understand correctly. Keep a `CameraPosition` whose center and rotation are updated when new GPS and compass events are received, and then when you want to apply the new camera use `CameraUpdateFactory.newCameraPosition` with your sensor-tracking `CameraPosition`. Does that make sense?

The new update behavior of only applying one update at a time is intentional. It's easier to reason about the camera behavior when there's at most one animation running at any given time.

**Comment by westnordost** [comment]
> It's easier to reason about the camera behavior when there's at most one animation running at any given time.

What do you mean? (_"[...] easier to reason [...]"_)

**Comment by matteblair** [comment]
It's easier for us to implement, which should lead to fewer bugs. I also think it's easier to mentally model as a user because the camera's behavior at any moment only depends on the most recent call to `updateCameraPosition`. That's the idea anyway.

**Comment by westnordost** [comment]
Well, you are right in that it is possible to create a wrapper around it so that not two threads manipulate camera parameters. Though it would have been nice if that wasn't necessary.

**Comment by westnordost** [comment]
I ended up with writing a convenience wrapper that takes over the whole camera animation: https://github.com/westnordost/StreetComplete/blob/new_tangram/app/src/main/java/de/westnordost/streetcomplete/map/tangram/CameraManager.kt

Most of the code there is actually only reimplementing the tangram animations with Android animators and wrapping it.

**The core is actually only a HashMap of property name** (rotation, tilt, zoom, position) **to currently running animation**. Then, whenever the camera is updated either via animation or immediately, the HashMap is checked for currently running animations on the properties that will be set/animated now and cancels only those that intersect.
This behavior makes it possible to set and animate different properties (at different speeds etc).

Finally, I changed the interface so that it is possible to write
```kotlin
cameraManager.updateCamera(500) {
    zoom = 18.5f
    tiltBy = 0.3f
}
```
(This is something like the Kotlin builder pattern, it Java it would look like
```java
CameraUpdate cameraUpdate = new CameraUpdate.Builder().zoom(18.5f).tiltBy(0.3f).build()
cameraManager.updateCamera(500, cameraUpdate)
```
or similar)

---

Of course it would be nice to see this (or a similar) easier and more flexible interface upstream.

The reason why I didn't make my convenience wrapper return to the old 0.9.6 interface where every single property can be set directly individually is because sometimes one might want to animate several properties together and also have them cancelled together if another intersecting camera update is made.
For example it should zoom to a certain position (zoom + pos) but tapping the "follow my GPS location" button should cancel that whole animation and not only the animation to the position.

---

In case you want to reconsider the current behavior, I can open a fresh ticket for that with my suggestion how the interface would ideally look.
